File: repo_tool/core/summary.py
import asyncio
import cProfile
import os
import pstats
from dataclasses import dataclass, field
from pathlib import Path
from pstats import SortKey
from typing import Any, Dict, List, Optional, TypeVar

import aiofiles
import tiktoken
from jinja2 import Environment, FileSystemLoader

from repo_tool.core.contants import DIGEST_DIR

# 型変数の定義
T = TypeVar("T")

data_size = 20
precision = 2
BATCH_SIZE = 100
MAX_FILE_SIZE = 5000
enable_profiling = os.getenv("ENABLE_PROFILING", "false").lower() == "true" or False
encoding = tiktoken.get_encoding("o200k_base")


@dataclass
class FileInfo:
    """ファイル処理に必要な情報を保持するデータクラス"""

    file_path: Path
    repo_path: Path


@dataclass
class FileType:
    extension: str
    count: int
    tokens: int


@dataclass
class FileData:
    name: str
    path: str
    extension: str
    tokens: int


@dataclass
class FileStats:
    file_count: int
    total_size: float
    average_size: float
    max_size: float
    min_size: float
    context_length: int
    extension_tokens: List[FileType] = field(default_factory=list)
    file_data: List[FileData] = field(default_factory=list)


@dataclass
class Summary:
    repository: str
    total_files: int
    total_size_kb: float
    average_file_size_kb: float
    max_file_size_kb: float
    min_file_size_kb: float
    file_types: List[FileType]
    context_length: int
    file_data: List[FileData]

    def generate_report(self, data_size: int = 20) -> None:
        """
        HTMLレポートを生成して保存する
        """
        # Jinja2 環境を設定
        env = Environment(loader=FileSystemLoader("templates"))
        env.filters["format_number"] = format_number
        template = env.get_template("report.html")

        # ファイルタイプをトークン数でソート
        sorted_file_types = sorted(
            self.file_types, key=lambda x: x.tokens, reverse=True
        )[:data_size]

        html_content = template.render(
            repo_name=self.repository,
            summary=self,
            file_types_labels=[ft.extension for ft in sorted_file_types],
            file_types_data=[ft.tokens for ft in sorted_file_types],
            file_sizes_labels=[item.name for item in self.file_data[:data_size]],
            file_sizes_data=[item.tokens for item in self.file_data[:data_size]],
            file_sizes_paths=[item.path for item in self.file_data[:data_size]],
            all_files=self.file_data,
        )

        # HTMLレポートを保存
        report_path = f"digests/{self.repository}_report.html"
        with open(report_path, "w", encoding="utf-8") as f:
            f.write(html_content)
        print(f"Report saved to {report_path}")


# カスタムフィルターを定義
def format_number(value: int | float | str) -> str:
    """数値をカンマ区切りにフォーマット"""
    if isinstance(value, (int, float)):
        return f"{value:,}"  # カンマ区切り
    return value


def generate_summary(
    repo_path: Path,
    file_list: List[Path],
) -> Summary:
    """
    ファイル統計のサマリーレポートを生成する

    Args:
        repo_path: リポジトリのパス
        file_list: 処理対象のファイルリスト
        enable_profiling: プロファイリングを有効にするかどうか (デフォルト: False)
    """
    profiler = None
    if enable_profiling:
        profiler = cProfile.Profile()
        profiler.enable()

    if not os.path.exists(DIGEST_DIR):
        os.makedirs(DIGEST_DIR, exist_ok=True)

    file_infos = [FileInfo(Path(f), repo_path) for f in file_list]
    file_stats = asyncio.run(process_files(file_infos))

    summary = Summary(
        repository=repo_path.name,
        total_files=file_stats.file_count,
        total_size_kb=round(file_stats.total_size, precision),
        average_file_size_kb=round(file_stats.average_size, precision),
        max_file_size_kb=round(file_stats.max_size, precision),
        min_file_size_kb=round(file_stats.min_size, precision),
        file_types=file_stats.extension_tokens,
        context_length=file_stats.context_length,
        file_data=file_stats.file_data,
    )

    if enable_profiling and profiler:
        profiler.disable()
        profile_stats = pstats.Stats(profiler).sort_stats(SortKey.TIME)
        profile_stats.print_stats()

    return summary


async def process_files(file_infos: List[FileInfo]) -> FileStats:
    """
    全ファイルの非同期処理と集計を行う
    """
    # バッチサイズを設定
    BATCH_SIZE = 100

    extension_data: Dict[str, Dict[str, int]] = {}
    total_size = 0
    file_sizes = []
    context_length = 0
    processed_files = []
    file_data_list = []

    # バッチ処理を実装
    for i in range(0, len(file_infos), BATCH_SIZE):
        batch = file_infos[i : i + BATCH_SIZE]
        tasks = [process_single_file(file_info) for file_info in batch]
        results = await asyncio.gather(*tasks)

        for result in results:
            if result is None:
                continue

            processed_files.append(result["path"])
            total_size += result["size"]
            context_length += result["tokens"]
            file_sizes.append(result["size"])

            file_data_list.append(
                FileData(
                    name=Path(result["path"]).name,
                    path=result["path"],
                    extension=result["extension"],
                    tokens=result["tokens"],
                )
            )

            ext = result["extension"]
            if ext not in extension_data:
                extension_data[ext] = {"count": 0, "tokens": 0}
            extension_data[ext]["count"] += 1
            extension_data[ext]["tokens"] += result["tokens"]

    # ファイルデータをトークン数でソート
    file_data_list.sort(key=lambda x: x.tokens, reverse=True)

    extension_tokens = [
        FileType(extension=ext, count=data["count"], tokens=data["tokens"])
        for ext, data in extension_data.items()
    ]

    file_count = len(processed_files)
    return FileStats(
        file_count=file_count,
        total_size=total_size,
        average_size=total_size / file_count if file_count > 0 else 0,
        max_size=max(file_sizes, default=0),
        min_size=min(file_sizes, default=0),
        extension_tokens=extension_tokens,
        context_length=context_length,
        file_data=file_data_list,
    )


async def process_single_file(file_info: FileInfo) -> Optional[Dict[str, Any]]:
    """
    単一ファイルの非同期処理を行う補助関数
    """
    try:
        relative_path = str(file_info.file_path.relative_to(file_info.repo_path))

        # stat呼び出しを先に行う
        try:
            file_size = file_info.file_path.stat().st_size / 1024  # bytes to KB
        except Exception:
            return None

        # 大きすぎるファイルはスキップ
        if file_size > MAX_FILE_SIZE:  # 1MB以上のファイルはスキップ
            return {
                "path": relative_path,
                "size": file_size,
                "tokens": 0,
                "extension": file_info.file_path.suffix.lower() or "no_extension",
            }

        async with aiofiles.open(
            file_info.file_path, "r", encoding="utf-8", errors="ignore"
        ) as f:
            content = await f.read()

        # トークン化処理
        tokens = len(encoding.encode(content))

        return {
            "path": relative_path,
            "size": file_size,
            "tokens": tokens,
            "extension": file_info.file_path.suffix.lower() or "no_extension",
        }
    except Exception as e:
        print(f"Error processing file {file_info.file_path}: {e}")
        return None

File: repo_tool_rust/src/summary.rs
use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use tiktoken_rs::get_bpe_from_model;
use tiktoken_rs::CoreBPE;
use tokio::fs;

// Constants
const DATA_SIZE: usize = 20;
const PRECISION: usize = 2;
const BATCH_SIZE: usize = 100;
const MAX_FILE_SIZE: u64 = 5000; // KB

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileType {
    pub extension: String,
    pub count: usize,
    pub tokens: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileData {
    pub name: String,
    pub path: String,
    pub extension: String,
    pub tokens: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FileStats {
    pub file_count: usize,
    pub total_size: f64,
    pub average_size: f64,
    pub max_size: f64,
    pub min_size: f64,
    pub context_length: usize,
    pub extension_tokens: Vec<FileType>,
    pub file_data: Vec<FileData>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Summary {
    pub repository: String,
    pub total_files: usize,
    pub total_size_kb: f64,
    pub average_file_size_kb: f64,
    pub max_file_size_kb: f64,
    pub min_file_size_kb: f64,
    pub file_types: Vec<FileType>,
    pub context_length: usize,
    pub file_data: Vec<FileData>,
}

pub struct FileInfo {
    file_path: PathBuf,
    repo_path: PathBuf,
}

pub async fn generate_summary(repo_path: &Path, file_list: &[PathBuf]) -> Result<Summary> {
    let file_infos: Vec<FileInfo> = file_list
        .iter()
        .map(|f| FileInfo {
            file_path: f.to_path_buf(),
            repo_path: repo_path.to_path_buf(),
        })
        .collect();

    let tokenizer = get_bpe_from_model("gpt-4o-2024-05-13")?;
    let file_stats = process_files(&file_infos, &tokenizer).await?;

    Ok(Summary {
        repository: repo_path
            .file_name()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string(),
        total_files: file_stats.file_count,
        total_size_kb: (file_stats.total_size * 100.0).round() / 100.0,
        average_file_size_kb: (file_stats.average_size * 100.0).round() / 100.0,
        max_file_size_kb: (file_stats.max_size * 100.0).round() / 100.0,
        min_file_size_kb: (file_stats.min_size * 100.0).round() / 100.0,
        file_types: file_stats.extension_tokens,
        context_length: file_stats.context_length,
        file_data: file_stats.file_data,
    })
}

async fn process_files(file_infos: &[FileInfo], tokenizer: &CoreBPE) -> Result<FileStats> {
    let mut extension_data: HashMap<String, (usize, usize)> = HashMap::new(); // (count, tokens)
    let mut total_size = 0.0;
    let mut file_sizes = Vec::new();
    let mut context_length = 0;
    let mut file_data_list = Vec::new();

    for chunk in file_infos.chunks(BATCH_SIZE) {
        let mut futures = Vec::new();
        for file_info in chunk {
            futures.push(process_single_file(file_info, tokenizer));
        }

        let results = futures::future::join_all(futures).await;

        for result in results {
            if let Ok(Some(file_result)) = result {
                total_size += file_result.size;
                context_length += file_result.tokens;
                file_sizes.push(file_result.size);

                file_data_list.push(FileData {
                    name: Path::new(&file_result.path)
                        .file_name()
                        .unwrap_or_default()
                        .to_string_lossy()
                        .to_string(),
                    path: file_result.path,
                    extension: file_result.extension.clone(),
                    tokens: file_result.tokens,
                });
                let entry = extension_data
                    .entry(file_result.extension.clone())
                    .or_insert((0, 0));
                entry.0 += 1;
                entry.1 += file_result.tokens;
            }
        }
    }

    // Sort file_data_list by tokens
    file_data_list.sort_by(|a, b| b.tokens.cmp(&a.tokens));

    let extension_tokens: Vec<FileType> = extension_data
        .into_iter()
        .map(|(ext, (count, tokens))| FileType {
            extension: ext,
            count,
            tokens,
        })
        .collect();

    let file_count = file_data_list.len();
    let average_size = if file_count > 0 {
        total_size / file_count as f64
    } else {
        0.0
    };

    Ok(FileStats {
        file_count,
        total_size,
        average_size,
        max_size: file_sizes.iter().fold(0.0, |a, &b| a.max(b)),
        min_size: file_sizes.iter().fold(f64::INFINITY, |a, &b| a.min(b)),
        context_length,
        extension_tokens,
        file_data: file_data_list,
    })
}

#[derive(Debug, Clone)]
struct FileResult {
    path: String,
    size: f64,
    tokens: usize,
    extension: String,
}

async fn process_single_file(file_info: &FileInfo, bpe: &CoreBPE) -> Result<Option<FileResult>> {
    let relative_path = file_info
        .file_path
        .strip_prefix(&file_info.repo_path)?
        .to_string_lossy()
        .to_string();

    let metadata = fs::metadata(&file_info.file_path).await?;
    let file_size = metadata.len() as f64 / 1024.0; // Convert to KB

    if file_size > MAX_FILE_SIZE as f64 {
        return Ok(Some(FileResult {
            path: relative_path,
            size: file_size,
            tokens: 0,
            extension: file_info
                .file_path
                .extension()
                .and_then(|ext| ext.to_str())
                .unwrap_or("no_extension")
                .to_lowercase(),
        }));
    }

    let content = fs::read_to_string(&file_info.file_path).await?;
    let tokens = bpe.encode_with_special_tokens(&content).len();

    Ok(Some(FileResult {
        path: relative_path,
        size: file_size,
        tokens,
        extension: file_info
            .file_path
            .extension()
            .and_then(|ext| ext.to_str())
            .unwrap_or("no_extension")
            .to_lowercase(),
    }))
}
